src/compiler.ts

```diff
@@ -7233,16 +7346,71 @@ export class Compiler extends DiagnosticEmitter {
   // <reference-counting>
 
   /** Makes a retain call, retaining the expression's value. */
-  makeRetain(expr: ExpressionRef): ExpressionRef {
+  makeRetain(
+    expr: ExpressionRef,
+    type: Type | null = null
+  ): ExpressionRef {
+    var module = this.module;
+
     var retainInstance = this.program.retainInstance;
     this.compileFunction(retainInstance);
-    return this.module.call(retainInstance.internalName, [ expr ], this.options.nativeSizeType);
+    if (type && type.isFunction) {
+      var exprLocal = this.currentFlow.getTempLocal(type)
+      var functionRetainCall = module.block(null, [
+        module.local_set(exprLocal.index, expr),
+        module.drop(
+          module.call(
+            retainInstance.internalName,
+            [
+              this.ifClosure(
+                module.local_get(exprLocal.index, type.toNativeType()),
+                this.getClosurePtr(module.local_get(exprLocal.index, type.toNativeType())),
+                module.i32(0)
+              )
+            ],
+            this.options.nativeSizeType
+          )
+        ),
+        module.local_get(exprLocal.index, type.toNativeType())
```

Comment:
From my understanding after talking to dcode: retains don't change the pointer, but it is expected that a retainment is an assignment of the form `x = retain(x)`. Before this change this wasn't a problem- because what was passed into retain was always exactly `x`. Now, what is passed into retain could be `x << 4` or `0` or `x`, so we just drop the retain result and do a get on the local.

Seems like that ought to be the same here, but I'll have to make sure that this isn't preventing ARC optimization

Reply:
The important aspect is that the optimize-arc pass sees that "`x` holds a reference" on `x = __retain(...)` (only), so it can eliminate the `__retain` when it sees a matching `__release(x)`. It completely ignores all other `__retain`s, and one must be very careful in these cases because the optimizer will still try to find another matching `x = __retain(...)`, potentially performing elimination of unbalanced branches or otherwise, leading to leaky or broken final code.

Reply:
Not super adept with binaryen but would it be possible to change what we consider the `full retain pattern` to `local.set(x, block(drop(_retain(..)), local.get(x))` to match what we have here? Seems like we'd just need to change the binaryen pass right? Since functionally these two approaches are the same given that `_retain()` is a no-op, we just need to change what our binaryen pass it looking for?

Seems like it would just be some minor changes to https://github.com/WebAssembly/binaryen/blob/e4d1e203c8bf8577e645dbf7ab265d1ec1c68bee/src/passes/PostAssemblyScript.cpp#L98

If not, we might have to reconsider the current strategy of using a retain(0) in anonymous function cases, since we're unable to reconstruct the function reference for the writeback

Reply:
As a last resort, perhaps. I still hope that there a ways to make it work without looking for more complicated patterns, though. Something along the lines of the following maybe

```ts
var x = __retain(
  isClosure(ptr)
    ? ptr << 4
    : 0
);
...
__release(x);
```

```ts
var x;
if (isClosure(ptr)) {
  x = __retain(ptr << 4);
} else {
  x = __retain(0);
}
...
__release(x);
```

Reply:
That could work but might be tricky to implement, because now we need to keep track of `x` and `ptr` so we know what to release later. I'll look into it